compiler/package.go

```diff
@@ -295,16 +214,74 @@ func Compile(srcs sources.Sources, importContext *ImportContext, minify bool) (_
 
 	return &Archive{
 		ImportPath:   srcs.ImportPath,
-		Name:         typesPkg.Name(),
+		Name:         srcs.Package.Name(),
 		Imports:      importedPaths,
-		Package:      typesPkg,
+		Package:      srcs.Package,
 		Declarations: allDecls,
 		FileSet:      srcs.FileSet,
 		Minified:     minify,
-		GoLinknames:  goLinknames,
+		GoLinknames:  srcs.GoLinknames,
 	}, nil
 }
 
+// PrepareAllSources prepares all sources for compilation by
+// parsing go linknames, type checking, sorting, simplifying, and
+// performing cross package analysis.
+// The results are stored in the provided sources.
+//
+// All sources must be given at the same time for cross package analysis to
+// work correctly. For consistency, the sources should be sorted by import path.
+func PrepareAllSources(allSources []*sources.Sources, importer sources.Importer, tContext *types.Context) error {
```

Comment:
This is the big change. In prior tickets I changed the code so we'd get all the Sources for all the packages prior to doing compile. This is all the preparation that was done in compile, instead being run on all the Sources at the same time prior to any package compiling.

To prevent this method from having tons of maps to store all the data per package, I just changed Sources to have the data stored in them and changes the Sources methods to populate those data field instead of returning the data. I also broke up some of those methods as needed to get all the parts prepared in the correct order.